backend/associations/serializers/election.py

```diff
@@ -1,83 +1,97 @@
+        )
+
+    def update(self, instance, validated_data):
+        raise NotImplementedError(
+            "This serializer cannot be used for database operations."
+        )
+
+
+class VoterSerializer(serializers.ModelSerializer):
+    user = serializers.PrimaryKeyRelatedField(
+        queryset=User.objects.all(), many=False, read_only=False
+    )
+
     class Meta:
-        model = Choice
-        read_only_fields = ("id", "name")
-        fields = read_only_fields
+        model = Voter
+        read_only_fields = ("id",)
+        fields = read_only_fields + ("user", "status", "election")
 
+    def to_representation(self, voter: Voter):
+        response = super(VoterSerializer, self).to_representation(voter)
+        response["user"] = UserShortSerializer(voter.user).data
+        return response
 
-class ChoiceSerializer(serializers.ModelSerializer):
-    election = serializers.PrimaryKeyRelatedField(read_only=True)
 
+class ChoiceSerializer(serializers.ModelSerializer):
     class Meta:
         model = Choice
-        read_only_fields = (
-            "id",
-            "election",
-        )  # 'election' will be provided by the user of the serializer.
-        fields = read_only_fields + ("name",)
+        read_only_fields = ("id",)
+        fields = read_only_fields + ("election", "name", "number_of_offline_votes")
 
 
-class BallotSerializer(serializers.ModelSerializer):
-    election = serializers.PrimaryKeyRelatedField(read_only=True)
-    choices = serializers.PrimaryKeyRelatedField(
-        queryset=Choice.objects.all(), many=True, read_only=False
+class ElectionVoterSerializer(serializers.ModelSerializer):
+    user = serializers.PrimaryKeyRelatedField(
+        queryset=User.objects.all(), many=False, read_only=False
     )
 
     class Meta:
-        model = Ballot
-        read_only_fields = (
-            "id",
-            "election",
-        )  # 'election' will be provided by the user of the serializer.
-        fields = read_only_fields + ("choices",)
-
+        model = Voter
+        fields = ("user", "status")
 
-class BallotShortSerializer(serializers.ModelSerializer):
-    election = serializers.PrimaryKeyRelatedField(read_only=True)
-    choices = serializers.PrimaryKeyRelatedField(many=True, read_only=True)
 
+class ElectionChoiceSerializer(serializers.ModelSerializer):
     class Meta:
-        model = Ballot
-        read_only_fields = ("id", "election", "choices")
-        fields = read_only_fields
+        model = Choice
+        fields = ("name", "number_of_offline_votes")
 
+    def to_representation(self, instance: Choice):
+        response = super(ElectionChoiceSerializer, self).to_representation(instance)
 
-class ElectionSerializer(serializers.ModelSerializer):
-    """This serializer gives access to the registered_voters and choices fields.
+        if instance.election.show_results:
+            response["number_of_online_votes"] = instance.number_of_online_votes
```

Comment:
Why always show `number_of_offline_votes` but sometimes `number_of_online_votes` ?

Reply:
Because nobody should see the results of the election until `show_results` is True, including the election administrator
However, the election administrator can always edit / see `number_of_offline_votes`